p2p/pkg/rpc/validator/service.go

```diff
@@ -0,0 +1,192 @@
+package validatorapi
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log/slog"
+	"net/http"
+	"strconv"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	validatorapiv1 "github.com/primev/mev-commit/p2p/gen/go/validatorapi/v1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type ValidatorRouterContract interface {
+	AreValidatorsOptedIn(valBLSPubKeys [][]byte) ([]bool, error)
+}
+
+type OptsGetter func(context.Context) (*bind.CallOpts, error)
+
+type Service struct {
+	validatorapiv1.UnimplementedValidatorServer
+	apiURL          string
+	validatorRouter ValidatorRouterContract
+	logger          *slog.Logger
+	metrics         *metrics
+}
+
+func NewService(apiURL string, validatorRouter ValidatorRouterContract, logger *slog.Logger) *Service {
+	return &Service{
+		apiURL:          apiURL,
+		validatorRouter: validatorRouter,
+		logger:          logger,
+		metrics:         newMetrics(),
+	}
+}
+
+type FinalityCheckpointsResponse struct {
+	Data struct {
+		PreviousJustified struct {
+			Epoch string `json:"epoch"`
+		} `json:"previous_justified"`
+		CurrentJustified struct {
+			Epoch string `json:"epoch"`
+		} `json:"current_justified"`
+		Finalized struct {
+			Epoch string `json:"epoch"`
+		} `json:"finalized"`
+	} `json:"data"`
+}
+
+type ProposerDutiesResponse struct {
+	Data []struct {
+		Pubkey string `json:"pubkey"`
+		Slot   string `json:"slot"`
+	} `json:"data"`
+}
+
+func (s *Service) GetEpoch(
+	ctx context.Context,
+	_ *validatorapiv1.EmptyMessage,
+) (*validatorapiv1.EpochResponse, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, s.apiURL+"/eth/v1/beacon/states/head/finality_checkpoints", nil)
+	if err != nil {
+		s.logger.Error("creating request", "error", err)
+		return nil, status.Errorf(codes.Internal, "creating request: %v", err)
+	}
+
+	req.Header.Set("accept", "application/json")
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		s.logger.Error("making request", "error", err)
+		return nil, status.Errorf(codes.Internal, "making request: %v", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		s.logger.Error("unexpected status code", "status", resp.StatusCode)
+		return nil, status.Errorf(codes.Internal, "unexpected status code: %v", resp.StatusCode)
+	}
+
+	var checkpointsResp FinalityCheckpointsResponse
+	if err := json.NewDecoder(resp.Body).Decode(&checkpointsResp); err != nil {
+		s.logger.Error("decoding response", "error", err)
+		return nil, status.Errorf(codes.Internal, "decoding response: %v", err)
+	}
+
+	currentEpoch, err := strconv.ParseUint(checkpointsResp.Data.CurrentJustified.Epoch, 10, 64)
```

Comment:
To avoid naming confusion

```suggestion
	currentjustifiedEpoch, err := strconv.ParseUint(checkpointsResp.Data.CurrentJustified.Epoch, 10, 64)
	currentEpoch := currentJustifiedEpoch + 1
```